File: frontend/js/views/second_income.js
// Second-income worth-it — net household benefit after taxes, childcare, and
// work costs, via /calc/second-income. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const pct = (n) => Number(n).toLocaleString(undefined, { maximumFractionDigits: 1 }) + '%';

export async function renderSecondIncome(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.secondinc.h1.title">// SECOND INCOME</span></h1>
        <p class="muted small" data-i18n="view.secondinc.hint.intro">
            What a household actually keeps from a second earner's income. It's taxed at the
            household's marginal rate (it stacks on top of the first), then childcare and
            work-related costs come out — what's left is the real benefit of working. Updates as
            you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.secondinc.h2.inputs">The second job</h2>
            <form id="secondinc-form" class="inline-form">
                <label><span data-i18n="view.secondinc.label.income">Second annual income ($)</span>
                    <input type="number" step="0.01" min="0" name="second_annual_income_usd" value="50000" required></label>
                <label><span data-i18n="view.secondinc.label.tax">Marginal tax rate (%)</span>
                    <input type="number" step="0.1" min="0" name="marginal_tax_rate_pct" value="30" required></label>
                <label><span data-i18n="view.secondinc.label.childcare">Annual childcare ($)</span>
                    <input type="number" step="0.01" min="0" name="annual_childcare_usd" value="18000"></label>
                <label><span data-i18n="view.secondinc.label.commute">Annual commute ($)</span>
                    <input type="number" step="0.01" min="0" name="annual_commute_usd" value="3000"></label>
                <label><span data-i18n="view.secondinc.label.other">Other work expenses ($)</span>
                    <input type="number" step="0.01" min="0" name="annual_other_work_expenses_usd" value="2000"></label>
            </form>
        </div>
        <div id="secondinc-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#secondinc-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            second_annual_income_usd: Number(fd.get('second_annual_income_usd')) || 0,
            marginal_tax_rate_pct: Number(fd.get('marginal_tax_rate_pct')) || 0,
            annual_childcare_usd: Number(fd.get('annual_childcare_usd')) || 0,
            annual_commute_usd: Number(fd.get('annual_commute_usd')) || 0,
            annual_other_work_expenses_usd: Number(fd.get('annual_other_work_expenses_usd')) || 0,
        };
        try {
            const r = await api.calcSecondIncome(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.secondinc.toast.error'), { level: 'error' });
        }
    };
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#secondinc-result');
    const cls = r.worth_it ? 'pos' : 'neg';
    const verdictKey = r.worth_it ? 'view.secondinc.verdict.worth' : 'view.secondinc.verdict.not';
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.secondinc.h2.result">The real benefit</h2>
            <div class="cards">
                <div class="card ${cls}"><div class="label" data-i18n="view.secondinc.card.net">Net benefit</div>
                    <div class="value ${cls}">${money(r.net_benefit_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.secondinc.card.keep">Keep rate</div>
                    <div class="value">${pct(r.keep_rate_pct)}</div></div>
                <div class="card ${cls}"><div class="label" data-i18n="view.secondinc.card.verdict">Verdict</div>
                    <div class="value ${cls}" data-i18n="${verdictKey}">—</div></div>
            </div>
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.secondinc.row.taxes">Taxes</td><td>${money(r.taxes_usd)}</td></tr>
                    <tr><td data-i18n="view.secondinc.row.aftertax">After-tax income</td><td>${money(r.after_tax_income_usd)}</td></tr>
                    <tr><td data-i18n="view.secondinc.row.costs">Total work costs</td><td>${money(r.total_costs_usd)}</td></tr>
                    <tr><td data-i18n="view.secondinc.row.keep">Keep rate</td><td>${pct(r.keep_rate_pct)}</td></tr>
                    <tr class="emph ${cls}"><td data-i18n="view.secondinc.row.net">Net benefit</td><td>${money(r.net_benefit_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/second_income.rs
//! Second-income worth-it analysis — what a household actually keeps from a
//! second earner's income after taxes, childcare, and work-related costs.
//!
//! The second income is taxed at the household's marginal rate (it stacks on
//! top of the first), then childcare and commuting/work costs come out:
//!
//! ```text
//! after-tax  = income × (1 − marginal rate)
//! net benefit = after-tax − childcare − commute − other work costs
//! keep rate   = net benefit / gross income
//! ```

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct SecondIncomeInput {
    pub second_annual_income_usd: f64,
    /// Combined marginal rate (federal + state + FICA) on the second income.
    pub marginal_tax_rate_pct: f64,
    #[serde(default)]
    pub annual_childcare_usd: f64,
    #[serde(default)]
    pub annual_commute_usd: f64,
    #[serde(default)]
    pub annual_other_work_expenses_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct SecondIncomeResult {
    pub taxes_usd: f64,
    pub after_tax_income_usd: f64,
    /// Childcare + commute + other work expenses.
    pub total_costs_usd: f64,
    /// After-tax income less all costs (what the household actually gains).
    pub net_benefit_usd: f64,
    /// Net benefit as a percent of gross second income.
    pub keep_rate_pct: f64,
    /// Whether the second income leaves the household ahead.
    pub worth_it: bool,
}

pub fn analyze(input: &SecondIncomeInput) -> SecondIncomeResult {
    let taxes = input.second_annual_income_usd * input.marginal_tax_rate_pct / 100.0;
    let after_tax = input.second_annual_income_usd - taxes;
    let costs =
        input.annual_childcare_usd + input.annual_commute_usd + input.annual_other_work_expenses_usd;
    let net = after_tax - costs;

    SecondIncomeResult {
        taxes_usd: taxes,
        after_tax_income_usd: after_tax,
        total_costs_usd: costs,
        net_benefit_usd: net,
        keep_rate_pct: if input.second_annual_income_usd > 0.0 {
            net / input.second_annual_income_usd * 100.0
        } else {
            0.0
        },
        worth_it: net > 0.0,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-6
    }

    fn base() -> SecondIncomeInput {
        SecondIncomeInput {
            second_annual_income_usd: 50_000.0,
            marginal_tax_rate_pct: 30.0,
            annual_childcare_usd: 18_000.0,
            annual_commute_usd: 3_000.0,
            annual_other_work_expenses_usd: 2_000.0,
        }
    }

    #[test]
    fn taxes_and_after_tax() {
        let r = analyze(&base());
        assert!(close(r.taxes_usd, 15_000.0));
        assert!(close(r.after_tax_income_usd, 35_000.0));
    }

    #[test]
    fn total_costs() {
        assert!(close(analyze(&base()).total_costs_usd, 23_000.0));
    }

    #[test]
    fn net_benefit() {
        // 35,000 − 23,000 = 12,000.
        assert!(close(analyze(&base()).net_benefit_usd, 12_000.0));
    }

    #[test]
    fn keep_rate() {
        // 12,000 / 50,000 = 24%.
        assert!(close(analyze(&base()).keep_rate_pct, 24.0));
    }

    #[test]
    fn worth_it_when_positive() {
        assert!(analyze(&base()).worth_it);
    }

    #[test]
    fn high_childcare_makes_it_not_worth_it() {
        let r = analyze(&SecondIncomeInput {
            annual_childcare_usd: 35_000.0,
            ..base()
        });
        assert!(r.net_benefit_usd < 0.0);
        assert!(!r.worth_it);
    }

    #[test]
    fn higher_tax_lowers_benefit() {
        let low = analyze(&base());
        let high = analyze(&SecondIncomeInput {
            marginal_tax_rate_pct: 45.0,
            ..base()
        });
        assert!(high.net_benefit_usd < low.net_benefit_usd);
    }

    #[test]
    fn no_costs_keeps_full_after_tax() {
        let r = analyze(&SecondIncomeInput {
            annual_childcare_usd: 0.0,
            annual_commute_usd: 0.0,
            annual_other_work_expenses_usd: 0.0,
            ..base()
        });
        assert!(close(r.net_benefit_usd, r.after_tax_income_usd));
        assert!(close(r.keep_rate_pct, 70.0));
    }
}
